File: android/app/src/main/java/com/volleycut/nativeanalysis/SuppressionPolicyEngine.java
package com.volleycut.nativeanalysis;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

/** Language-neutral millisecond policy algebra from the suppression product contract. */
final class SuppressionPolicyEngine {
    enum Policy {
        NONE("none", "No suppression", "none", 0, 0),
        CONSERVATIVE("conservative", "Conservative", "zero-non-exempt-misses", 2_000, 500),
        BALANCED("balanced", "Balanced", "aggressive-intermediate", 1_500, 500),
        AGGRESSIVE("aggressive", "Aggressive", "raw-connected", 0, 0);

        final String wireName;
        final String label;
        final String recordedPolicy;
        final long agreementPaddingMs;
        final long agreementJoinMs;

        Policy(
                String wireName, String label, String recordedPolicy,
                long agreementPaddingMs, long agreementJoinMs
        ) {
            this.wireName = wireName;
            this.label = label;
            this.recordedPolicy = recordedPolicy;
            this.agreementPaddingMs = agreementPaddingMs;
            this.agreementJoinMs = agreementJoinMs;
        }

        static Policy fromWireName(String value) {
            for (Policy policy : values()) if (policy.wireName.equals(value)) return policy;
            return NONE;
        }
    }

    private enum Source { ALL_LABELS_V2, PREVIOUS_PRODUCTION }
    private record Tagged(
            String id, long startMs, long endMs, float confidence, Source source,
            long paddedStartMs, long paddedEndMs
    ) {}
    private record Span(long startMs, long endMs) {}
    private record PolicySpan(long startMs, long endMs, Policy policy) {}

    private SuppressionPolicyEngine() {}

    static AnalysisTypes.SuppressionAnalysis build(
            List<AnalysisTypes.Interval> allLabelsV2,
            List<AnalysisTypes.Interval> previousProduction,
            float[] probabilities,
            List<AnalysisTypes.Interval> decoded,
            double durationSeconds
    ) {
        long durationMs = milliseconds(durationSeconds);
        List<Tagged> raw = tagged(allLabelsV2, previousProduction, durationMs);
        ArrayList<PolicySpan> policySpans = new ArrayList<>();
        for (Policy policy : List.of(Policy.CONSERVATIVE, Policy.BALANCED, Policy.AGGRESSIVE)) {
            for (Span span : eligibleRawSpans(raw, policy, durationMs)) {
                policySpans.add(new PolicySpan(span.startMs, span.endMs, policy));
            }
        }

        ArrayList<AnalysisTypes.SuppressionSuggestion> suggestions = new ArrayList<>();
        for (int decodedIndex = 0; decodedIndex < decoded.size(); decodedIndex++) {
            AnalysisTypes.Interval event = decoded.get(decodedIndex);
            long eventStart = Math.max(0, milliseconds(event.start()));
            long eventEnd = Math.min(durationMs, milliseconds(event.end()));
            if (eventEnd <= eventStart) continue;
            List<Tagged> eventSources = raw.stream()
                    .filter(item -> overlaps(eventStart, eventEnd, item.startMs, item.endMs))
                    .toList();
            if (eventSources.isEmpty()) continue;
            String logicalId = logicalId(eventStart, eventEnd, eventSources);
            LinkedHashSet<Long> boundaries = new LinkedHashSet<>();
            boundaries.add(eventStart);
            boundaries.add(eventEnd);
            for (PolicySpan candidate : policySpans) {
                long start = Math.max(eventStart, candidate.startMs);
                long end = Math.min(eventEnd, candidate.endMs);
                if (end > start) {
                    boundaries.add(start);
                    boundaries.add(end);
                }
            }
            List<Long> ordered = boundaries.stream().sorted().toList();
            ArrayList<AnalysisTypes.SuppressionSuggestion> atoms = new ArrayList<>();
            for (int index = 0; index + 1 < ordered.size(); index++) {
                long start = ordered.get(index);
                long end = ordered.get(index + 1);
                if (end <= start) continue;
                long midpoint = start + (end - start) / 2;
                List<String> eligible = List.of(Policy.CONSERVATIVE, Policy.BALANCED, Policy.AGGRESSIVE)
                        .stream().filter(policy -> policySpans.stream().anyMatch(span ->
                                span.policy == policy && midpoint >= span.startMs && midpoint < span.endMs
                        )).map(policy -> policy.wireName).toList();
                if (eligible.isEmpty()) continue;
                List<String> sourceIds = eventSources.stream()
                        .filter(item -> overlaps(start, end, item.startMs, item.endMs))
                        .map(Tagged::id).sorted().toList();
                if (sourceIds.isEmpty()) continue;
                atoms.add(new AnalysisTypes.SuppressionSuggestion(
                        logicalId,
                        logicalId + ":" + start + ":" + end,
                        start,
                        end,
                        event.confidence(),
                        sourceIds,
                        eligible
                ));
            }
            for (AnalysisTypes.SuppressionSuggestion atom : atoms) {
                AnalysisTypes.SuppressionSuggestion previous = suggestions.isEmpty()
                        ? null : suggestions.get(suggestions.size() - 1);
                if (previous != null && previous.logicalId().equals(atom.logicalId())
                        && previous.endMs() == atom.startMs()
                        && previous.eligiblePolicyIds().equals(atom.eligiblePolicyIds())) {
                    ArrayList<String> ids = new ArrayList<>(previous.sourceProductionIds());
                    atom.sourceProductionIds().forEach(id -> { if (!ids.contains(id)) ids.add(id); });
                    ids.sort(String::compareTo);
                    suggestions.set(suggestions.size() - 1, new AnalysisTypes.SuppressionSuggestion(
                            previous.logicalId(),
                            previous.logicalId() + ":" + previous.startMs() + ":" + atom.endMs(),
                            previous.startMs(), atom.endMs(),
                            Math.max(previous.score(), atom.score()),
                            List.copyOf(ids), previous.eligiblePolicyIds()
                    ));
                } else suggestions.add(atom);
            }
        }
        return new AnalysisTypes.SuppressionAnalysis(
                FeatureSchema.SUPPRESSION_MODEL_ID,
                FeatureSchema.SUPPRESSION_ARTIFACT_SHA256,
                FeatureSchema.SUPPRESSION_WEIGHTS_SHA256,
                FeatureSchema.SUPPRESSION_DECODER_VERSION,
                probabilities.clone(),
                List.copyOf(decoded),
                List.copyOf(suggestions)
        );
    }

    static List<AnalysisTypes.SuppressionSuggestion> active(
            AnalysisTypes.SuppressionAnalysis analysis, Policy policy
    ) {
        if (analysis == null || policy == Policy.NONE) return List.of();
        return analysis.suggestions().stream()
                .filter(item -> item.eligiblePolicyIds().contains(policy.wireName))
                .sorted(Comparator.comparingLong(AnalysisTypes.SuppressionSuggestion::startMs)
                        .thenComparingLong(AnalysisTypes.SuppressionSuggestion::endMs))
                .toList();
    }

    private static List<Tagged> tagged(
            List<AnalysisTypes.Interval> allLabelsV2,
            List<AnalysisTypes.Interval> previousProduction,
            long durationMs
    ) {
        ArrayList<Tagged> result = new ArrayList<>();
        addTagged(result, allLabelsV2, Source.ALL_LABELS_V2, "all-labels-v2", durationMs);
        addTagged(result, previousProduction, Source.PREVIOUS_PRODUCTION,
                "previous-production", durationMs);
        return List.copyOf(result);
    }

    private static void addTagged(
            List<Tagged> target, List<AnalysisTypes.Interval> source, Source kind,
            String prefix, long durationMs
    ) {
        for (int index = 0; index < source.size(); index++) {
            AnalysisTypes.Interval interval = source.get(index);
            long start = Math.max(0, milliseconds(interval.start()));
            long end = Math.min(durationMs, milliseconds(interval.end()));
            if (end <= start) continue;
            target.add(new Tagged(
                    prefix + ":" + String.format("%04d", index + 1) + ":" + start + ":" + end,
                    start, end, interval.confidence(), kind, start, end
            ));
        }
    }

    private static List<Span> eligibleRawSpans(List<Tagged> raw, Policy policy, long durationMs) {
        List<Tagged> padded = raw.stream().map(item -> new Tagged(
                item.id, item.startMs, item.endMs, item.confidence, item.source,
                Math.max(0, item.startMs - policy.agreementPaddingMs),
                Math.min(durationMs, item.endMs + policy.agreementPaddingMs)
        )).sorted(Comparator.comparingLong(Tagged::paddedStartMs)
                .thenComparingLong(Tagged::paddedEndMs)).toList();
        ArrayList<List<Tagged>> components = new ArrayList<>();
        long componentEnd = Long.MIN_VALUE;
        for (Tagged item : padded) {
            long gap = item.paddedStartMs - componentEnd;
            if (components.isEmpty() || (gap > 0 && gap >= policy.agreementJoinMs)) {
                components.add(new ArrayList<>());
                componentEnd = item.paddedEndMs;
            } else componentEnd = Math.max(componentEnd, item.paddedEndMs);
            components.get(components.size() - 1).add(item);
        }
        ArrayList<Span> eligible = new ArrayList<>();
        for (List<Tagged> component : components) {
            Set<Source> sources = new HashSet<>();
            component.forEach(item -> sources.add(item.source));
            if (sources.size() != 1) continue;
            component.forEach(item -> eligible.add(new Span(item.startMs, item.endMs)));
        }
        return merge(eligible);
    }

    private static List<Span> merge(List<Span> source) {
        ArrayList<Span> merged = new ArrayList<>();
        source.stream().sorted(Comparator.comparingLong(Span::startMs)
                .thenComparingLong(Span::endMs)).forEach(item -> {
            Span previous = merged.isEmpty() ? null : merged.get(merged.size() - 1);
            if (previous == null || item.startMs > previous.endMs) merged.add(item);
            else merged.set(merged.size() - 1,
                    new Span(previous.startMs, Math.max(previous.endMs, item.endMs)));
        });
        return List.copyOf(merged);
    }

    private static String logicalId(long start, long end, List<Tagged> source) {
        String canonical = start + ":" + end + ":" + source.stream()
                .map(Tagged::id).sorted().reduce((left, right) -> left + "," + right).orElse("");
        try {
            byte[] digest = MessageDigest.getInstance("SHA-256")
                    .digest(canonical.getBytes(StandardCharsets.UTF_8));
            StringBuilder value = new StringBuilder("S-");
            for (int index = 0; index < 10; index++) value.append(String.format("%02x", digest[index]));
            return value.toString();
        } catch (NoSuchAlgorithmException error) {
            throw new IllegalStateException(error);
        }
    }

    private static boolean overlaps(long leftStart, long leftEnd, long rightStart, long rightEnd) {
        return leftStart < rightEnd && rightStart < leftEnd;
    }

    private static long milliseconds(double seconds) {
        return Math.round(seconds * 1_000.0);
    }
}

File: ios/Sources/VolleyCore/SuppressionPolicyEngine.swift
import CryptoKit
import Foundation

public struct SuppressionSuggestion: Codable, Equatable, Sendable {
    public var logicalId: String, fragmentId: String
    public var startMs: Int64, endMs: Int64
    public var score: Float
    public var sourceProductionIds: [String], eligiblePolicyIds: [String]
    public init(logicalId: String, fragmentId: String, startMs: Int64, endMs: Int64, score: Float,
                sourceProductionIds: [String], eligiblePolicyIds: [String]) {
        self.logicalId = logicalId; self.fragmentId = fragmentId; self.startMs = startMs; self.endMs = endMs
        self.score = score; self.sourceProductionIds = sourceProductionIds; self.eligiblePolicyIds = eligiblePolicyIds
    }
    /// Preserves logical identity across policy fragments for the existing draft override algebra.
    public var region: SuppressionRegion {
        .init(id: fragmentId, logicalId: logicalId, startMs: startMs, endMs: endMs, eligiblePolicyIds: eligiblePolicyIds)
    }
}
public struct SuppressionAnalysis: Codable, Equatable, Sendable {
    public let modelId: String, artifactSha256: String, weightsSha256: String, decoderVersion: String
    public let probabilities: [Float], decodedIntervals: [Interval], suggestions: [SuppressionSuggestion]
}

/// Android suppression-policy-v1 millisecond algebra. Suggestions describe eligible regions;
/// EditorMath applies a draft's explicit decisions and whole-rally/veto-region scope.
public enum SuppressionPolicyEngine {
    public static let contractVersion = "suppression-policy-v1"
    public enum Policy: String, Codable, Sendable, CaseIterable {
        case none, conservative, balanced, aggressive
        public static func fromWireName(_ value: String) -> Policy { Policy(rawValue: value) ?? .none }
        public var label: String {
            switch self { case .none: "No suppression"; case .conservative: "Conservative"; case .balanced: "Balanced"; case .aggressive: "Aggressive" }
        }
        public var recordedPolicy: String {
            switch self { case .none: "none"; case .conservative: "zero-non-exempt-misses"; case .balanced: "aggressive-intermediate"; case .aggressive: "raw-connected" }
        }
        fileprivate var padding: Int64 { self == .conservative ? 2000 : self == .balanced ? 1500 : 0 }
        fileprivate var join: Int64 { self == .conservative || self == .balanced ? 500 : 0 }
    }
    private struct Tagged {
        let id: String, start: Int64, end: Int64, allLabels: Bool
        var paddedStart: Int64, paddedEnd: Int64
    }
    private struct Span { let start: Int64, end: Int64 }
    private static let policies: [Policy] = [.conservative, .balanced, .aggressive]
    private static func milliseconds(_ seconds: Double) -> Int64 {
        // Java Math.round ties towards positive infinity. Callers validate representability first.
        Int64(floor(seconds * 1000 + 0.5))
    }
    private static func overlaps(_ a: Int64, _ b: Int64, _ c: Int64, _ d: Int64) -> Bool { a < d && c < b }

    public static func build(allLabelsV2: [Interval], previousProduction: [Interval], probabilities: [Float],
                             decoded: [Interval], duration: Double) throws -> SuppressionAnalysis {
        guard duration.isFinite, duration >= 0, duration < Double(Int64.max / 2000), probabilities.allSatisfy(\.isFinite),
              (allLabelsV2 + previousProduction + decoded).allSatisfy({ $0.start.isFinite && $0.end.isFinite &&
                  abs($0.start) < Double(Int64.max / 2000) && abs($0.end) < Double(Int64.max / 2000) && $0.confidence.isFinite }) else {
            throw AnalysisError.invalid("Invalid suppression policy intervals")
        }
        let durationMs = milliseconds(duration)
        var raw: [Tagged] = []
        for (source, allLabels, prefix) in [(allLabelsV2, true, "all-labels-v2"), (previousProduction, false, "previous-production")] {
            for (index, interval) in source.enumerated() {
                let start = max(0, milliseconds(interval.start)), end = min(durationMs, milliseconds(interval.end))
                if end <= start { continue }
                let id = prefix + ":" + String(format: "%04d", index + 1) + ":\(start):\(end)"
                raw.append(Tagged(id: id, start: start, end: end, allLabels: allLabels, paddedStart: start, paddedEnd: end))
            }
        }
        let policySpans: [(Span, Policy)] = policies.flatMap { policy in
            eligible(raw, policy, durationMs).map { ($0, policy) }
        }
        var suggestions: [SuppressionSuggestion] = []
        for event in decoded {
            let eventStart = max(0, milliseconds(event.start)), eventEnd = min(durationMs, milliseconds(event.end))
            if eventEnd <= eventStart { continue }
            let sources = raw.filter { overlaps(eventStart, eventEnd, $0.start, $0.end) }
            if sources.isEmpty { continue }
            let canonical = "\(eventStart):\(eventEnd):" + sources.map(\.id).sorted().joined(separator: ",")
            let logicalId = "S-" + SHA256.hash(data: Data(canonical.utf8)).prefix(10).map { String(format: "%02x", $0) }.joined()
            var boundaries: Set<Int64> = [eventStart, eventEnd]
            for (span, _) in policySpans {
                let start = max(eventStart, span.start), end = min(eventEnd, span.end)
                if end > start { boundaries.insert(start); boundaries.insert(end) }
            }
            let ordered = boundaries.sorted()
            for index in 0..<(ordered.count - 1) {
                let start = ordered[index], end = ordered[index + 1], midpoint = start + (end - start) / 2
                let eligiblePolicies: [Policy] = policies.filter { policy in
                    policySpans.contains { span, candidatePolicy in
                        candidatePolicy == policy && midpoint >= span.start && midpoint < span.end
                    }
                }
                let eligibleIds: [String] = eligiblePolicies.map(\.rawValue)
                if eligibleIds.isEmpty { continue }
                let ids = sources.filter { overlaps(start, end, $0.start, $0.end) }.map(\.id).sorted()
                if ids.isEmpty { continue }
                if let previous = suggestions.last, previous.logicalId == logicalId, previous.endMs == start, previous.eligiblePolicyIds == eligibleIds {
                    suggestions[suggestions.count - 1] = SuppressionSuggestion(logicalId: logicalId, fragmentId: "\(logicalId):\(previous.startMs):\(end)",
                        startMs: previous.startMs, endMs: end, score: max(previous.score, event.confidence),
                        sourceProductionIds: Array(Set(previous.sourceProductionIds + ids)).sorted(), eligiblePolicyIds: eligibleIds)
                } else {
                    suggestions.append(SuppressionSuggestion(logicalId: logicalId, fragmentId: "\(logicalId):\(start):\(end)",
                        startMs: start, endMs: end, score: event.confidence, sourceProductionIds: ids, eligiblePolicyIds: eligibleIds))
                }
            }
        }
        return SuppressionAnalysis(modelId: SuppressionModelRunner.modelId, artifactSha256: SuppressionModelRunner.artifactSha256,
            weightsSha256: SuppressionModelRunner.weightsSha256, decoderVersion: SuppressionModelRunner.decoderVersion,
            probabilities: probabilities, decodedIntervals: decoded, suggestions: suggestions)
    }

    public static func active(_ analysis: SuppressionAnalysis?, policy: Policy) -> [SuppressionSuggestion] {
        guard let analysis, policy != .none else { return [] }
        return analysis.suggestions.filter { $0.eligiblePolicyIds.contains(policy.rawValue) }.sorted { ($0.startMs, $0.endMs) < ($1.startMs, $1.endMs) }
    }

    private static func eligible(_ raw: [Tagged], _ policy: Policy, _ duration: Int64) -> [Span] {
        let padded = raw.map { item in
            var item = item; item.paddedStart = max(0, item.start - policy.padding); item.paddedEnd = min(duration, item.end + policy.padding); return item
        }.sorted { ($0.paddedStart, $0.paddedEnd) < ($1.paddedStart, $1.paddedEnd) }
        var components: [[Tagged]] = [], componentEnd: Int64 = 0
        for item in padded {
            let gap = item.paddedStart - componentEnd
            if components.isEmpty || (gap > 0 && gap >= policy.join) {
                components.append([]); componentEnd = item.paddedEnd
            } else { componentEnd = max(componentEnd, item.paddedEnd) }
            components[components.count - 1].append(item)
        }
        let singleSourceComponents: [[Tagged]] = components.filter { Set($0.map(\.allLabels)).count == 1 }
        let eligibleItems: [Tagged] = singleSourceComponents.flatMap { $0 }
        let spans: [Span] = eligibleItems.map { Span(start: $0.start, end: $0.end) }
        let source: [Span] = spans.sorted { ($0.start, $0.end) < ($1.start, $1.end) }
        var merged: [Span] = []
        for item in source {
            if let previous = merged.last, item.start <= previous.end {
                merged[merged.count - 1] = Span(start: previous.start, end: max(previous.end, item.end))
            } else { merged.append(item) }
        }
        return merged
    }
}
